### DISCUSSION TITLE
kong.service.set_target() not work in tcp plugin for dynamic proxy， and kong not have balancer() callback for tcp plugin

### DISCUSSION THREAD

**Discussion by stanstanwang** [Kong/kong#9849] [Ideas and feature requests]
### Discussed in https://github.com/Kong/kong/discussions/9809

<div type='discussions-op-text'>

<sup>Originally posted by **stanstanwang** November 23, 2022</sup>
i write a dynamic tcp proxy plugin,   my logic is like this, my kong version is 3.0.x

``` lua
function plugin:preread(plugin_conf)
  -- your custom code here, for example;
  kong.log.inspect('preread', plugin_conf)
  kong.service.set_target('host.docker.internal', 9206);
end
```

but got an error: 
```
2022/11/23 03:29:23 [error] 22323#0: *73628 [kong] init.lua:290 [myplugin] /usr/local/share/lua/5.1/kong/pdk/private/phases.lua:90: function cannot be called in preread phase (only in: access) while prereading client data, client: 1
92.168.48.1, server: 0.0.0.0:7000
```
anyone can help
</div>

**Comment by stanstanwang** [comment]
@bungle @hishamhm  I'm sorry to bother you，can you help me, thanks

I temporarily solve the problem by ngx.ctx, but this need edit the source code,  

`/usr/local/share/lua/5.1/kong/init.lua`
``` lua

  -- [warning] temporarily  add custom tcp dynamic proxy logic
  if is_stream_module and ngx.ctx.stream_balancer_data then
    balancer_data.ip = ngx.ctx.stream_balancer_data.ip
    balancer_data.port = ngx.ctx.stream_balancer_data.port
  end
  -- [warning] end
  
  current_try.ip   = balancer_data.ip
  current_try.port = balancer_data.port
  
  -- set the targets as resolved
  ngx_log(ngx_DEBUG, "setting address (try ", balancer_data.try_count, "): ",
    balancer_data.ip, ":", balancer_data.port)

```

can tcp plugin add a lifecycle  `balancer_by_lua` ,  we need this to add dynamic proxy logic, like the http `acccess`  callback, thanks ~

**Comment by StarlightIbuki** [comment]
Thanks for reporting. Currently, we do not support interventions with the balancer for TCP plugins. This is missing from the doc of `kong.service.set_target()`, and we will fix this.

And the support requires further discussion. We are on it.
